dialects : add comb dialect

dshaaban01 opened the pull request:
This is the comb dialect from circuit and the corresponding tests.

PapyChacal commented on tests/filecheck/dialects/comb/comb_ops.mlir:
```diff
@@ -0,0 +1,69 @@
+  %icmp = "comb.icmp"(%lhsi1, %rhsi1) {"predicate" = 2 : index, "two_state"} : (i1, i1) -> i1
+  // CHECK-NEXT: %icmp = "comb.icmp"(%lhsi1, %rhsi1) {"predicate" = 2 : index, "two_state"} : (i1, i1) -> i1
+
+  }) : () -> ()
```
Missing EOL :slightly_smiling_face:

superlopuh replied:
```suggestion
  }) : () -> ()

```

PapyChacal commented on xdsl/dialects/comb.py:
```diff
@@ -0,0 +1,245 @@
+class VariadicCombOp(IRDLOperation):
+    """
+    A variadic comb operation. It has a variadic number of operands, and a single
+    result, all of the same type.
```
```suggestion
    result, all of the same integer type.
```

superlopuh commented on xdsl/dialects/riscv.py:
looks like the diff in this file accidentally got into this branch

superlopuh commented on xdsl/dialects/comb.py:
```diff
@@ -0,0 +1,245 @@
+    inputs: VarOperand = var_operand_def(T)
+    result: OpResult = result_def(T)
+
+    two_state: UnitAttr | None = opt_attr_def(UnitAttr)
```
Can you please add documentation for this variable?

dshaaban01 replied:
Which one? The two_state variable?

superlopuh replied:
yes, please, it's not obvious to me just by the name what it's supposed to do

PapyChacal reviewed (commented):
Thanks!

Left some minor nitpick comments

I feel like the RISC-V canonicalization should be another PR? We usually try to split independent changes in individual PRs.
But the code seems sound and the testing is done, so I wouldn't block merging here.

superlopuh commented on xdsl/dialects/comb.py:
```diff
@@ -0,0 +1,245 @@
+    two_state: UnitAttr | None = opt_attr_def(UnitAttr)
+
+
+class VariadicCombOp(IRDLOperation):
```
```suggestion
class VariadicCombOperation(IRDLOperation, abc.ABC):
```

We've been annotating abstract base operation classes explicitly with `abc.ABC`, and appending `Operation` as opposed to `Op` as a convention. It's not documented anywhere and I'm not sure it's 100% consistent across all dialects but I find it pretty useful as a visual guide, could you please add it here also? This is more of a nit, so feel free to ignore.

superlopuh replied:
Can you also please add an `__init__` to the two base classes?

math-fehr replied:
We were thinking of adding the `__init__` after in this case?
So we can have more reasonable PRs ;)
Also, same things for the parser/printer

dshaaban01 replied:
Working on finishing the __init__ and parser/printer this morning :)

superlopuh replied:
I would recommend merging operations one by one but in finished form rather than leaving bits for later but I guess that's a subjective call. We merged the riscv dialect almost instruction by instruction and I thought that went quite well.

math-fehr replied:
I'm fine with both, so fine for me!

math-fehr replied:
Actually I'm not sure.
Since everything is already implemented now, can we just merge this and then add the __init__? The __init__ should not be a breaking change, and adding __init__ here would just make the PR so much bigger (since we would add the tests as well?)

superlopuh replied:
Sure that's not a problem

compor replied:
I was wondering about the "merging operations one by one" and I thought of suggesting it, but since now everything is in place, it rather impedes progress.

I've tried to look in git history how has this been handled in other dialects, but I've mostly stumbled in PRs that moved files around, so not sure what might be a good practice.

superlopuh commented on xdsl/dialects/comb.py:
```diff
@@ -0,0 +1,245 @@
+
+
+@irdl_op_definition
+class ICmpOp(IRDLOperation):
```
can you please add `__init__`s here and below?

tobiasgrosser commented:
Very nice. Thank you @dshaaban01

webmiche commented:
Similar to #1597, this is a mirrored dialect from circt right? Could add docu with a link to that dialect at the top of the file, please?

Also, this PR would benefit from circt interoperation too. :thinking: (let's keep the discussion in the other PR)

superlopuh commented on tests/filecheck/backend/riscv/canonicalize.mlir:
```diff
@@ -126,4 +129,9 @@ builtin.module {
 
 // CHECK-NEXT:   riscv.fsw %i2, %f2, 12 : (!riscv.reg<>, !riscv.freg<>) -> ()
 
+// CHECK-NEXT:   %add_lhs_rhs = riscv.li 2 : () -> !riscv.reg<>
```
I'm confused about this change, I thought we merged it in #1566 already? I wonder why GitHub is not showing merge conflicts here.

superlopuh replied:
Can you please merge main into this PR to get rid of this change?

tobiasgrosser commented:
I very much agree. And this is a takeawatfor myself. Especially for new contributors, I need to make sure that the first tasks clearly specify how they are broken into pieces. Or even better, that I follow my own advice and report on plans and discussions on zulip.

dshaaban01 commented:
I just had a long session fixing the entire PR mess that I created with Chris! I am so sorry about that guys!

I have now addressed all of the comments (except the addition of any init functions) in this PR and have now pushed them for your review. Chris has suggested that we then merge this PR and I continue my work on the other PR. 

To recap what the objective of the two PR's are:
This PR (dalia/risc-5-canonicalization): basic introduction of the dialect + tests
Other PR (dalia/init-comb): introduction of init, parser, printer functions for the base classes

math-fehr reviewed (approved):
That looks good to me, is that okay for you to merge it @superlopuh ?
So we can look at the next one

superlopuh commented on tests/filecheck/dialects/comb/comb_ops.mlir:
```diff
@@ -0,0 +1,69 @@
+  %icmp = "comb.icmp"(%lhsi1, %rhsi1) {"predicate" = 2 : index, "two_state"} : (i1, i1) -> i1
+  // CHECK-NEXT: %icmp = "comb.icmp"(%lhsi1, %rhsi1) {"predicate" = 2 : index, "two_state"} : (i1, i1) -> i1
+  }) : () -> ()
+  
```
```suggestion

```
some stray trailing whitespace

compor replied:
@dshaaban01 I believe the permanent solution to avoid spurious insertions in the future in VSCode is the setting

```json
"files.trimTrailingWhitespace": true
```

Searching for "trailing" in preferences should find the relevant UI checkbox.

math-fehr replied:
Hmm, I'm actually confused why this isn't checked by pre-commit, maybe I made a mistake there?

compor replied:
@math-fehr hmm, yeah that was my initial thought, but I didn't have time to look into it and `black`.

tobiasgrosser commented:
Yay!
